File: Contests/BiWeekly/27/3_course_schedule_iv.py
from collections import defaultdict


class Solution:
    def checkIfPrerequisite(self, n: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
        if not prerequisites:
            return [False]*len(queries)
        mapp = defaultdict(list)
        for p in prerequisites:
            mapp[p[1]].append(p[0])
        #as the map is changing in the loop
        c = list(mapp.keys()).copy()
        for k in c:
            l = set(mapp[k])
            for i in mapp[k]:
                l = l.union(set(mapp[i]))
            mapp[k] = list(l)
        #call this twice so as to update the map completely
        for k in c:
            l = set(mapp[k])
            for i in mapp[k]:
                l = l.union(set(mapp[i]))
            mapp[k] = list(l)
        res = []
        for i in queries:
            if i[0] in mapp[i[1]]:
                res.append(True)
            else:
                res.append(False)
        return res

File: Contests/BiWeekly/27/3_course_schedule_iv.cpp
class Solution {
public:
    vector<bool> checkIfPrerequisite(int n, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        bool isConnected[n][n];
        for(int i = 0; i < n; i++){
            for(int j = 0; j < n; j++){
                isConnected[i][j] = false;
            }
        }
        for (auto &p : prerequisites)
            isConnected[p[0]][p[1]] = true;
        
        //floyd warshall --> 1e6
        for (int k = 0; k < n; k++)
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++)
                    isConnected[i][j] = isConnected[i][j] || isConnected[i][k] && connected[k][j];
        
        vector<bool> ans;
        for (auto &q : queries)
            ans.push_back(connected[q[0]][q[1]]);
        return ans;
    }
};